From fa3509ed08449b95d9826748da10e51f7873610a Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:51:35 -0400
Subject: [iommu] vt-d: Only call dmar_acpi_dev_scope_init() if DRHD units present

Message-id: <20140905015135.31297.82673.stgit@gir.stowe>
Patchwork-id: 91351
O-Subject: [RHEL7.1 PATCH 58/65] iommu/vt-d: Only call dmar_acpi_dev_scope_init() if DRHD units present
Bugzilla: 1129880 1087643
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1087643
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129880
Upstream Status: 63b4262478ee6384c7efb5279be54236e5d6361b

commit 63b4262478ee6384c7efb5279be54236e5d6361b
Author: David Woodhouse <David.Woodhouse@intel.com>
Date:   Fri Mar 28 11:28:40 2014 +0000

    iommu/vt-d: Only call dmar_acpi_dev_scope_init() if DRHD units present

    As pointed out by Joerg and fixed in commit 11f1a7768 ("iommu/vt-d: Check
    for NULL pointer in dmar_acpi_dev_scope_init(), this code path can
    bizarrely get exercised even on AMD IOMMU systems with IRQ remapping
    enabled.

    In addition to the defensive check for NULL which Joerg added, let's also
    just avoid calling the function at all if there aren't an Intel IOMMU
    units in the system.

    Signed-off-by: David Woodhouse <David.Woodhouse@intel.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/dmar.c b/drivers/iommu/dmar.c
index d3d1dca..e74d77b 100644
--- a/drivers/iommu/dmar.c
+++ b/drivers/iommu/dmar.c
@@ -698,13 +698,13 @@ int __init dmar_dev_scope_init(void)
  if (dmar_dev_scope_status != 1)
   return dmar_dev_scope_status;
 
- dmar_acpi_dev_scope_init();
-
  if (list_empty(&dmar_drhd_units)) {
   dmar_dev_scope_status = -ENODEV;
  } else {
   dmar_dev_scope_status = 0;
 
+  dmar_acpi_dev_scope_init();
+
   for_each_pci_dev(dev) {
    if (dev->is_virtfn)
     continue;
-- 
1.7.1